File: mip_reference/dotnet-aspnet/Mip/Models/CogsRequest.cs
using System.Text.Json.Serialization;

namespace Mip.Models;

/// <summary>
/// Tracks Certificate of Good Standing requests (inbound and outbound).
/// </summary>
public class CogsRequest
{
    public static readonly string[] Statuses = { "PENDING", "APPROVED", "DECLINED" };

    public string SharedIdentifier { get; set; } = Guid.NewGuid().ToString();
    public string Direction { get; set; } = ""; // "inbound" or "outbound"
    public string TargetMipIdentifier { get; set; } = "";
    public string TargetOrg { get; set; } = "";
    public Dictionary<string, string?> RequestingMember { get; set; } = new();
    public string RequestedMemberNumber { get; set; } = "";
    public string? Notes { get; set; }
    public string Status { get; set; } = "PENDING";
    public Dictionary<string, object?>? Certificate { get; set; }
    public string? DeclineReason { get; set; }
    public string CreatedAt { get; set; } = DateTime.UtcNow.ToString("o");

    [JsonIgnore]
    public bool IsPending => Status == "PENDING";
    [JsonIgnore]
    public bool IsApproved => Status == "APPROVED";
    [JsonIgnore]
    public bool IsDeclined => Status == "DECLINED";
    [JsonIgnore]
    public bool IsInbound => Direction == "inbound";
    [JsonIgnore]
    public bool IsOutbound => Direction == "outbound";

    public void Approve(Member member, Dictionary<string, object?> issuingOrg)
    {
        Status = "APPROVED";
        Certificate = new Dictionary<string, object?>
        {
            ["shared_identifier"] = SharedIdentifier,
            ["status"] = "APPROVED",
            ["good_standing"] = member.GoodStanding,
            ["issued_at"] = DateTime.UtcNow.ToString("o"),
            ["valid_until"] = DateTime.UtcNow.AddDays(90).ToString("o"),
            ["issuing_organization"] = issuingOrg,
            ["member_profile"] = member.ToMemberProfile()
        };
    }

    public void Decline(string? reason = null)
    {
        Status = "DECLINED";
        DeclineReason = reason;
        Certificate = new Dictionary<string, object?>
        {
            ["shared_identifier"] = SharedIdentifier,
            ["status"] = "DECLINED",
            ["good_standing"] = false,
            ["reason"] = reason
        };
    }

    public Dictionary<string, object?> ToRequestPayload()
    {
        var payload = new Dictionary<string, object?>
        {
            ["shared_identifier"] = SharedIdentifier,
            ["requesting_member"] = RequestingMember,
            ["requested_member_number"] = RequestedMemberNumber
        };

        if (!string.IsNullOrEmpty(Notes))
            payload["notes"] = Notes;

        return payload;
    }

    public Dictionary<string, object?> ToReplyPayload()
    {
        return Certificate ?? new Dictionary<string, object?>
        {
            ["shared_identifier"] = SharedIdentifier,
            ["status"] = Status
        };
    }

    /// <summary>
    /// Create from received request payload.
    /// </summary>
    public static CogsRequest FromRequest(Dictionary<string, object?> payload, string senderMipId, string senderOrg)
    {
        var requestingMember = new Dictionary<string, string?>();
        if (payload.TryGetValue("requesting_member", out var rm) && rm is Dictionary<string, object?> rmDict)
        {
            foreach (var kvp in rmDict)
            {
                requestingMember[kvp.Key] = kvp.Value?.ToString();
            }
        }

        return new CogsRequest
        {
            SharedIdentifier = payload.GetValueOrDefault("shared_identifier")?.ToString() ?? Guid.NewGuid().ToString(),
            Direction = "inbound",
            TargetMipIdentifier = senderMipId,
            TargetOrg = senderOrg,
            RequestingMember = requestingMember,
            RequestedMemberNumber = payload.GetValueOrDefault("requested_member_number")?.ToString() ?? "",
            Notes = payload.GetValueOrDefault("notes")?.ToString()
        };
    }
}

File: mip-ruby/lib/member_interchange_protocol/models/cogs_request.rb
# frozen_string_literal: true

module MemberInterchangeProtocol
  module Models
    # Tracks Certificate of Good Standing requests (inbound and outbound)
    class CogsRequest
      STATUSES = %w[PENDING APPROVED DECLINED].freeze

      attr_accessor :status, :certificate, :decline_reason
      attr_reader :shared_identifier, :direction, :target_mip_identifier,
                  :target_org, :requesting_member, :requested_member_number,
                  :notes, :created_at

      def initialize(attrs = {})
        @shared_identifier = attrs[:shared_identifier] || SecureRandom.uuid
        @direction = attrs[:direction] # 'inbound' or 'outbound'
        @target_mip_identifier = attrs[:target_mip_identifier]
        @target_org = attrs[:target_org]
        @requesting_member = attrs[:requesting_member] || {}
        @requested_member_number = attrs[:requested_member_number]
        @notes = attrs[:notes]
        @status = attrs.fetch(:status, 'PENDING')
        @certificate = attrs[:certificate]
        @decline_reason = attrs[:decline_reason]
        @created_at = attrs.fetch(:created_at, Time.now.iso8601)
      end

      def pending?
        @status == 'PENDING'
      end

      def approved?
        @status == 'APPROVED'
      end

      def declined?
        @status == 'DECLINED'
      end

      def inbound?
        @direction == 'inbound'
      end

      def outbound?
        @direction == 'outbound'
      end

      def approve!(member, issuing_org)
        @status = 'APPROVED'
        @certificate = {
          shared_identifier: @shared_identifier,
          status: 'APPROVED',
          good_standing: member.good_standing,
          issued_at: Time.now.iso8601,
          valid_until: (Time.now + (90 * 24 * 60 * 60)).iso8601, # 90 days
          issuing_organization: issuing_org,
          member_profile: member.to_member_profile
        }
      end

      def decline!(reason = nil)
        @status = 'DECLINED'
        @decline_reason = reason
        @certificate = {
          shared_identifier: @shared_identifier,
          status: 'DECLINED',
          good_standing: false,
          reason: reason
        }
      end

      def to_request_payload
        {
          shared_identifier: @shared_identifier,
          requesting_member: @requesting_member,
          requested_member_number: @requested_member_number,
          notes: @notes
        }.compact
      end

      def to_reply_payload
        @certificate || {
          shared_identifier: @shared_identifier,
          status: @status
        }
      end

      # Create from received request payload
      def self.from_request(payload, sender_mip_id, sender_org)
        new(
          shared_identifier: payload['shared_identifier'] || SecureRandom.uuid,
          direction: 'inbound',
          target_mip_identifier: sender_mip_id,
          target_org: sender_org,
          requesting_member: payload['requesting_member'] || {},
          requested_member_number: payload['requested_member_number'],
          notes: payload['notes']
        )
      end
    end
  end
end
